datafusion/core/src/dataframe/mod.rs

```diff
@@ -1271,8 +1273,13 @@ impl DataFrame {
     /// ```
     pub async fn cache(self) -> Result<DataFrame> {
         let context = SessionContext::new_with_state(self.session_state.clone());
+        // type cast
```

Comment:
Rather than explcilty running (only) coercion (which is also run as part of execution) I suggest you use the schema that comes directly from the output stream

Something like (copied from `Self::collect`): https://docs.rs/datafusion/latest/src/datafusion/dataframe/mod.rs.html#756-760
```
        let task_ctx = Arc::new(self.task_ctx());
        let plan = self.create_physical_plan().await?;
        let schema = plan.schema();
        collect(plan, task_ctx).await
```

Reply:
Thanks for your advice